Ensure that `copy` and `getindex` maintain `parent` field

spurll opened the pull request:
Closes #199

spurll committed:
Ensure that `copy` also copies the `parent` field

Closes #199

spurll commented:
Hm, I think that Codecov report is wrong, since I didn't touch most of those impacted files.

However, I realise that I'll need to make another commit to solve an outstanding issue related to sorting `DataTables`, so I'm going to mark this WiP.

ararslan commented:
> Hm, I think that Codecov report is wrong

Yeah, coverage is weird.

> However, I realise that I'll need to make another commit to solve an outstanding issue related to sorting DataTables

Is that directly related to this change?

spurll commented:
The problem I'm currently trying to solve is the fact that the new `NullableArray` created by `getindex` doesn't maintain the parent reference either:

```julia
julia> str = "something"
"something"

julia> arr = [WeakRefString(pointer(str, 1), 4, 1), WeakRefString(pointer(str, 5), 5, 5)]
2-element Array{WeakRefString{UInt8},1}:
 "some" 
 "thing"

julia> na = NullableArray{WeakRefString{UInt8}, 1}(arr, [false, false], Vector{UInt8}(str))
2-element NullableArrays.NullableArray{WeakRefString{UInt8},1}:
 "some" 
 "thing"

julia> na.parent
9-element Array{UInt8,1}:
 0x73
 0x6f
 0x6d
 0x65
 0x74
 0x68
 0x69
 0x6e
 0x67

julia> na[:].parent
0-element Array{UInt8,1}
```

It's certainly related, though maybe it warrants a separate PR.

spurll committed:
Ensure that `getindex` maintains `parent` field

Calls to `similar(X::NullableArray, ...)` now return a `NullableArray`
with the same `parent` field as `X`, allowing `getindex` (and `sort`,
etc.) to maintain a reference to the parent memory block where
necessary.

Also improves constructor support for `parent` field.

spurll commented:
This should now be good to go. Though it's possible that the change I made to `similar` will be contentious, it is necessary to maintain the parent reference without massive changes to the codebase.

nalimilan reviewed (approved):
Thanks, this makes sense, though that `parent` field is really an ugly hack... I'd like a check by @quinnj, who introduced this feature.

nalimilan commented on test/indexing.jl:
```diff
@@ -57,6 +57,12 @@ module TestIndexing
+    # ensure that the NullableArray returned by getindex maintains its parent field
+    str = "something"
+    x = NullableArray([pointer(str), pointer(str, 5)], [false, false], Vector{UInt8}(str))
+    @test isequal(x[1:2].parent, x.parent)
```
Use `===` Same below..

spurll replied:
Thanks. I've made that change, and the corresponding change to test/primitives.jl as well.

nalimilan reviewed (changes requested):
Woops, I had missed some details.

spurll committed:
Fix identity checks for parent field in tests

quinnj reviewed (approved):
This looks fine. I have new code in WeakRefStrings that introduces a WeakRefStringArray so that we don't need this kind of `parent` field anymore. I'm going to try and do new releases for all of DataStreams this week that uses it.
